File: macos/Sources/Clawix/AppState/UserFacingFailure.swift
import Foundation
import OSLog

enum UserFacingFailureKind: String, Equatable {
    case backendUnavailable
    case daemonUnavailable
    case permissionDenied
    case modelUnavailable
    case networkOffline
    case serviceUnavailable
    case unknown
}

struct UserFacingFailure: Equatable {
    let kind: UserFacingFailureKind
    let rawMessage: String
    let retryable: Bool

    private static let logger = Logger(
        subsystem: Bundle.main.bundleIdentifier ?? "com.example.clawix",
        category: "failure"
    )

    static func classify(_ message: String) -> UserFacingFailure {
        let trimmed = message.trimmingCharacters(in: .whitespacesAndNewlines)
        let lower = trimmed.lowercased()

        if containsAny(lower, [
            "not connected to the internet",
            "internet connection appears to be offline",
            "network connection lost",
            "cannot connect to host",
            "cannot find host",
            "timed out",
            "offline"
        ]) {
            return UserFacingFailure(kind: .networkOffline, rawMessage: trimmed, retryable: true)
        }

        if containsAny(lower, [
            "permission denied",
            "permissions were denied",
            "permission was denied",
            "workspace is denied",
            "workspacedenied",
            "not authorized",
            "unauthorized"
        ]) {
            return UserFacingFailure(kind: .permissionDenied, rawMessage: trimmed, retryable: true)
        }

        if containsAny(lower, [
            "model unavailable",
            "model is unavailable",
            "model not found",
            "unknown model",
            "not downloaded",
            "download a model",
            "no such model"
        ]) {
            return UserFacingFailure(kind: .modelUnavailable, rawMessage: trimmed, retryable: true)
        }

        if containsAny(lower, [
            "background bridge",
            "bridge is not ready",
            "daemonunreachable",
            "daemon unreachable",
            "daemon is unavailable",
            "reconnects",
            "reconnecting"
        ]) {
            return UserFacingFailure(kind: .daemonUnavailable, rawMessage: trimmed, retryable: true)
        }

        if containsAny(lower, [
            "backend binary not found",
            "backend app-server failed",
            "clawix backend is not running",
            "agent runtime is unavailable",
            "runtime is unavailable",
            "backend not ready"
        ]) {
            return UserFacingFailure(kind: .backendUnavailable, rawMessage: trimmed, retryable: true)
        }

        if containsAny(lower, [
            "service is unavailable",
            "service unavailable",
            "servicenotready",
            "service not ready",
            "could not reach"
        ]) {
            return UserFacingFailure(kind: .serviceUnavailable, rawMessage: trimmed, retryable: true)
        }

        return UserFacingFailure(kind: .unknown, rawMessage: trimmed, retryable: true)
    }

    var displayMessage: String {
        switch kind {
        case .backendUnavailable:
            return L10n.t("The agent runtime is unavailable. Try again after it reconnects.")
        case .daemonUnavailable:
            return L10n.t("The background bridge is unavailable. Try again after it reconnects.")
        case .permissionDenied:
            return L10n.t("Permission was denied. Review permissions, then try again.")
        case .modelUnavailable:
            return L10n.t("That model is not available. Pick another model and try again.")
        case .networkOffline:
            return L10n.t("The network appears to be offline. Reconnect, then try again.")
        case .serviceUnavailable:
            return L10n.t("The service is unavailable. Try again in a moment.")
        case .unknown:
            return rawMessage.isEmpty ? L10n.t("Request failed. Try again in a moment.") : rawMessage
        }
    }

    var chatLine: String {
        String(format: L10n.t("Error: %@"), displayMessage)
    }

    func log(surface: String) {
        Self.logger.warning(
            "failure surface=\(surface, privacy: .public) kind=\(kind.rawValue, privacy: .public) retryable=\(retryable, privacy: .public) detail=\(rawMessage, privacy: .private)"
        )
    }

    private static func containsAny(_ haystack: String, _ needles: [String]) -> Bool {
        needles.contains { haystack.contains($0) }
    }
}

File: web/src/lib/user-facing-failure.ts
import { t } from "../localization/i18n";

export type UserFacingFailureKind =
  | "backendUnavailable"
  | "daemonUnavailable"
  | "permissionDenied"
  | "modelUnavailable"
  | "networkOffline"
  | "serviceUnavailable"
  | "unknown";

export interface UserFacingFailure {
  kind: UserFacingFailureKind;
  message: string;
  retryable: boolean;
}

export function classifyUserFacingFailure(reason: string | null | undefined): UserFacingFailure {
  const raw = (reason ?? "").trim();
  const lower = raw.toLowerCase();

  if (hasAny(lower, [
    "not connected to the internet",
    "internet connection appears to be offline",
    "network connection lost",
    "network is unreachable",
    "cannot connect to host",
    "cannot find host",
    "timed out",
    "offline",
  ])) {
    return {
      kind: "networkOffline",
      message: t("The network appears to be offline. Reconnect, then try again."),
      retryable: true,
    };
  }

  if (hasAny(lower, [
    "permission denied",
    "forbidden",
    "http 401",
    "http 403",
    "invalid api key",
    "authentication failed",
    "auth-failed",
    "not authorized",
    "unauthorized",
  ])) {
    return {
      kind: "permissionDenied",
      message: t("Permission was denied. Review permissions, then try again."),
      retryable: true,
    };
  }

  if (hasAny(lower, [
    "model unavailable",
    "model is unavailable",
    "model not found",
    "unknown model",
    "no model available",
    "not downloaded",
    "download a model",
    "no such model",
  ])) {
    return {
      kind: "modelUnavailable",
      message: t("That model is not available. Pick another model and try again."),
      retryable: true,
    };
  }

  if (hasAny(lower, [
    "background bridge",
    "bridge is not ready",
    "daemonunreachable",
    "daemon unreachable",
    "daemon is unavailable",
    "closed code=",
    "dead-link",
    "open-failed",
  ])) {
    return {
      kind: "daemonUnavailable",
      message: t("The background bridge is unavailable. Try again after it reconnects."),
      retryable: true,
    };
  }

  if (hasAny(lower, [
    "agent runtime is unavailable",
    "runtime is unavailable",
    "backend not ready",
  ])) {
    return {
      kind: "backendUnavailable",
      message: t("The agent runtime is unavailable. Try again after it reconnects."),
      retryable: true,
    };
  }

  if (hasAny(lower, [
    "service unavailable",
    "service not ready",
    "could not reach",
  ])) {
    return {
      kind: "serviceUnavailable",
      message: t("The service is unavailable. Try again in a moment."),
      retryable: true,
    };
  }

  return {
    kind: "unknown",
    message: t("Request failed. Try again in a moment."),
    retryable: true,
  };
}

export function logUserFacingFailure(surface: string, failure: UserFacingFailure): void {
  console.warn("[Clawix.failure]", {
    surface,
    kind: failure.kind,
    retryable: failure.retryable,
  });
}

function hasAny(value: string, needles: string[]): boolean {
  return needles.some((needle) => value.includes(needle));
}
